### PR TITLE
feat: detangle chainId / domainId EVM assumptions

### PR DESCRIPTION
**Pull request by paulbalaji**
### Description

Detangle assumption that chainId == domainId for EVM chains.
- required to support new domain IDs for sept26 chain deploy batch https://github.com/hyperlane-xyz/hyperlane-registry/pull/352

Domain IDs and Chain Names are still unique, but chainId is no longer guaranteed to be a unique identifier.
- required to support shadow lumia (current deployment) + prod lumia (new deployment with different domain id)

Touches:
- evm core/ism/hook/warp modules
- multiprovider + chain metadata manager
- safe service
- renaming `ChainNameOrId` to `ChainNameOrDomain` (where bulk of the diff comes from)

### Drive-by changes

- ~~add name of a `chain` to `Annotated` types, since chain ID on a populated transaction is a looser tie to a chain than `domainId` or `chain`~~
- ~~simplify multiprovider.sendtransaction, since `chain` is provided with the annotated transaction~~

- ensure domainId is not optional in the chain metadata schema
	- requires registry fix https://github.com/hyperlane-xyz/hyperlane-registry/pull/357
- update ethers v5 json rpc submitter to require a `chain` as well

### Related issues

- mostly redoes work from https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/4599
- will have to be rechecked after zksync changes are also in

### Backward compatibility

no

### Testing

ci, manual

### CODE REVIEW

File: typescript/sdk/src/providers/transactions/transformer/ethersV5/EV5InterchainAccountTxTransformer.ts
Comment (tkporter) [inline typescript/sdk/src/providers/transactions/transformer/ethersV5/EV5InterchainAccountTxTransformer.ts:46]: does this need an assert on `chain`?

File: typescript/infra/scripts/funding/fund-deterministic-key-from-deployer.ts
Comment (tkporter) [inline typescript/infra/scripts/funding/fund-deterministic-key-from-deployer.ts:82]: I did kinda like the old interface fwiw but will defer to ppl closer to the SDK

**Review by tkporter — COMMENTED** [review]
lgtm but deferring to others

File: typescript/infra/scripts/funding/fund-deterministic-key-from-deployer.ts
Comment (paulbalaji) [reply typescript/infra/scripts/funding/fund-deterministic-key-from-deployer.ts:82]: fair, I just felt it was cleaner to cut the arg out since it's redundant to mention `chain` twice

File: typescript/sdk/src/metadata/ChainMetadataManager.ts
Comment (paulbalaji) [reply typescript/sdk/src/metadata/ChainMetadataManager.ts:68]: CLI still defaults to domainId being the chainId, so we'll still be able to catch duplicates in the general case - we can also add stricter tests in the registry itself rather than trying to shape it here

> don't create new deploys on chains that already have mailboxes

the cli currently checks if we've got a mailbox for the selected `chain`

File: typescript/sdk/src/providers/transactions/transformer/ethersV5/EV5InterchainAccountTxTransformer.ts
Comment (paulbalaji) [reply typescript/sdk/src/providers/transactions/transformer/ethersV5/EV5InterchainAccountTxTransformer.ts:46]: We don't *need* one because `AnnotatedEV5Transaction` extends from an ethers `PopulatedTransaction` where all fields are optional - so we need the assert to ensure values from that type are properly set.

```ts
import type { PopulatedTransaction as EV5Transaction } from 'ethers';

export type AnnotatedEV5Transaction = Annotated<EV5Transaction>;
```
image


Whereas `chain` is a required property of the `Annotated<>` wrapper we're using


```ts
export type Annotated<T> = T & {
  chain: string;
  annotation?: string;
};
```

File: typescript/infra/scripts/funding/fund-deterministic-key-from-deployer.ts
Comment (xeno097) [reply typescript/infra/scripts/funding/fund-deterministic-key-from-deployer.ts:82]: I prefer the cleaner approach in my case

File: typescript/sdk/src/providers/transactions/transformer/ethersV5/EV5InterchainAccountTxTransformer.ts
Comment (tkporter) [reply typescript/sdk/src/providers/transactions/transformer/ethersV5/EV5InterchainAccountTxTransformer.ts:46]: makes sense!

**Review by xeno097 — APPROVED** [review]
Good effort 🔥

File: typescript/sdk/src/providers/transactions/transformer/ethersV5/EV5InterchainAccountTxTransformer.ts
Comment (yorhodes) [reply typescript/sdk/src/providers/transactions/transformer/ethersV5/EV5InterchainAccountTxTransformer.ts:46]: `Pick` and `Required` might be useful here
https://www.typescriptlang.org/docs/handbook/utility-types.html#requiredtype

**Review by yorhodes — COMMENTED** [review]
submitting stale review

File: typescript/cli/examples/submit/strategy/json-rpc-chain-strategy.yaml
Comment (yorhodes) [inline typescript/cli/examples/submit/strategy/json-rpc-chain-strategy.yaml:4]: just flagging that the top level key is redundant with the inner key

File: typescript/cli/src/deploy/warp.ts
Comment (yorhodes) [inline typescript/cli/src/deploy/warp.ts:886]: the key of this is a chainId as a string?

File: typescript/sdk/src/types.ts
Comment (yorhodes) [inline typescript/sdk/src/types.ts:10]: I personally think changing this types name is not really necessary and is the main reason this diff is so large

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts
Comment (yorhodes) [inline typescript/infra/scripts/funding/fund-keys-from-deployer.ts:820]: isnt the multiprovider already EVM specific?

File: typescript/widgets/src/chains/ChainAddMenu.tsx
Comment (yorhodes) [inline typescript/widgets/src/chains/ChainAddMenu.tsx:160]: honestly im not sure we should remove this constraint from the multiprovider as a signer can think they are signing for chain X but their transaction can be replayed on chain Y

File: typescript/widgets/src/messages/useMessageStage.ts
Comment (yorhodes) [inline typescript/widgets/src/messages/useMessageStage.ts:180]: seems like an unnecessary diff

**Review by yorhodes — APPROVED** [review]
mostly lgtm
I dont think this needed to such a breaking change

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts
Comment (paulbalaji) [reply typescript/infra/scripts/funding/fund-keys-from-deployer.ts:820]: the multiprovider is evm specific but the function itself is from the chain metadata manager which is not chain specific

it's effectively a shorthand for ensuring we get a number in EVM land

File: typescript/cli/src/deploy/warp.ts
Comment (paulbalaji) [reply typescript/cli/src/deploy/warp.ts:886]: that's right - but it's because the lodash `groupBy` helper returns a Dictionary whose keys are always strings

File: typescript/cli/examples/submit/strategy/json-rpc-chain-strategy.yaml
Comment (paulbalaji) [reply typescript/cli/examples/submit/strategy/json-rpc-chain-strategy.yaml:4]: aye, just copying the pattern from the other submitter strategies here

File: typescript/widgets/src/chains/ChainAddMenu.tsx
Comment (paulbalaji) [reply typescript/widgets/src/chains/ChainAddMenu.tsx:160]: I think this is probably fine for us because my understanding is with our model we want to make sure that if we have a chain X and chain Y with the same chain ID, they are both actually chain X but just with different domain X and domain Y

File: typescript/cli/src/deploy/warp.ts
Comment (yorhodes) [reply typescript/cli/src/deploy/warp.ts:886]: I did not realize we added lodash
@jmrossy would have a conniption

Comment (paulbalaji) [reply typescript/cli/src/deploy/warp.ts:886]: aha yeah idk when it got added but it's been there for a little bit

Comment (yorhodes) [reply typescript/cli/src/deploy/warp.ts:886]: https://github.com/hyperlane-xyz/hyperlane-monorepo/commit/1474865ae14d014a86893d79ac0a36eea43e4a70 I guess he was onboard :)

Comment (paulbalaji) [reply typescript/cli/src/deploy/warp.ts:886]: I forgot I was the one who added it 🤣

File: typescript/sdk/src/providers/transactions/submitter/TxSubmitterInterface.ts
Comment (paulbalaji) [reply typescript/sdk/src/providers/transactions/submitter/TxSubmitterInterface.ts:25]: added to scratch notes for future

File: typescript/utils/src/types.ts
Comment (paulbalaji) [reply typescript/utils/src/types.ts:21]: reduces need to parseInt in places by making the expected type explicit in the evm-specific classes

File: typescript/cli/src/deploy/warp.ts
Comment (jmrossy) [reply typescript/cli/src/deploy/warp.ts:886]: Just be sure to use `lodash-es`!

**Pull request merged**
